drake/solvers/Optimization.h

```diff
@@ -114,6 +114,8 @@ namespace Drake {
     }
     virtual ~Constraint() {}
 
+    // TODO: consider using a Ref for `y` too.  This will require the client
+    // to do allocation, but also allows it to choose stack allocation instead.
     virtual void eval(const Eigen::Ref<const Eigen::VectorXd>& x, Eigen::VectorXd& y) const = 0;
     virtual void eval(const Eigen::Ref<const TaylorVecXd>& x, TaylorVecXd& y) const = 0;  // move this to DifferentiableConstraint derived class if/when we need to support non-differentiable functions
```

Comment:
just to make sure we're on the same page... this function should only be expected to compile if the input output relationship is DifferentiableAlmostEverywhere